### List of Papers

This table lists the papers from which our multi-tissue dataset is drawn. UID is intended to be a unique, stable ID within this project only, ie if any paper gets dropped from the set we are analysiing do not reuse it's UID for any new papers that we may include later. Ensure the UID remains suitable for a key field in a database table in case we use this approach.

| Paper UID | GH Issue# | DOI | PubMed |First Author | Year | Title |
|-----|-----------|-----|--------|-------------|------|-------|
|1|[#6](https://github.com/UCL/Mutein/issues/6)|[doi](https://doi.org/10.1158/2159-8290.cd-20-1092)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/33087317/)|Fowler|2021|Selection of oncogenic mutant clones in normal human skin varies with body site|
|2|[#6](https://github.com/UCL/Mutein/issues/6)|[doi](https://doi.org/10.1126/science.aau3879)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/30337457/)|Martincorena|2018|Somatic mutant clones colonize the human esophagus with age|
|3|[#6](https://github.com/UCL/Mutein/issues/6)|[doi](https://doi.org/10.1126/science.aaa6806)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/25999502/)|Martincorena|2015|High burden and pervasive positive selection of somatic mutations in normal human skin|
|4|[#8](https://github.com/UCL/Mutein/issues/8)|[doi](https://doi.org/10.1038/s41586-019-1672-7)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/31645730/)|Lee-Six|2019|The landscape of somatic mutation in normal colorectal epithelial cells|
|5|[#9](https://github.com/UCL/Mutein/issues/9)|[doi](https://doi.org/10.1038/s41467-018-06331-w)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/30323172/)|Keogh|2018|High prevalence of focal and multi-focal somatic genetic variants in the human brain|
|6|[#11](https://github.com/UCL/Mutein/issues/11)|[doi](https://doi.org/10.1073/pnas.1902510116)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/30992375/)|Zhang|2019|Single-cell whole-genome sequencing reveals the functional landscape of somatic mutations in B lymphocytes across the human lifespan|
|7|[#12](https://github.com/UCL/Mutein/issues/12)|[doi](https://doi.org/10.1038/s41586-018-0811-x)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/30602793/)|Yokoyama|2019|Age-related remodelling of oesophageal epithelia by mutated cancer drivers|
|8|[#13](https://github.com/UCL/Mutein/issues/13)|[doi](https://doi.org/10.1182/blood-2017-04-777029)|[pubmed](https://pubmed.ncbi.nlm.nih.gov/28655780/)|Buscarlet|2017|DNMT3A and TET2 dominate clonal hematopoiesis and demonstrate benign phenotypes and different genetic predispositions|
|9|||[pubmed](https://pubmed.ncbi.nlm.nih.gov/35177608/)|Yamaguchi|2022|Spatiotemporal dynamics of clonal selection and diversification in normal endometrial epithelium|
|10|||[pubmed](https://pubmed.ncbi.nlm.nih.gov/33004514/)|Lawson|2020|Extensive heterogeneity in somatic mutation and selection in the human bladder|
|11|||[pubmed](https://pubmed.ncbi.nlm.nih.gov/31645727/)|Brunner|2019|Somatic mutations and clonal dynamics in healthy and cirrhotic human liver|

### Summary of Datasets Mentioned in each Paper

This section maps out the references to datasets found in the papers. Where the same dataset is referred to in more than one paper or top-level accession number these are retained as redundant entries in the lists below. For the non-redundant list see the Dataset Accession list below.

- Paper 1 (Fowler 2021):
  - EGA Accessions:
    - EGAS00001001933
      - Study [EGAS00001001933](https://ega-archive.org/studies/EGAS00001001933)
      - Dataset [EGAD00001006969](https://ega-archive.org/datasets/EGAD00001006969)
      - Dataset [EGAD00001004158](https://ega-archive.org/datasets/EGAD00001004158)
      - Dataset [EGAD00001006194](https://ega-archive.org/datasets/EGAD00001006194)
    - EGAS00001002165
      - Study [EGAS00001002165](https://ega-archive.org/studies/EGAS00001002165)
      - Dataset [EGAD00001006194](https://ega-archive.org/datasets/EGAD00001006194)
    - EGAS00001002708
      - Study [EGAS00001002708](https://ega-archive.org/studies/EGAS00001002708)
      - Dataset [EGAD00001006194](https://ega-archive.org/datasets/EGAD00001006194)
    - EGAS00001002416
      - Study [EGAS00001002416](https://ega-archive.org/studies/EGAS00001002416)
      - Dataset [EGAD00001004159](https://ega-archive.org/datasets/EGAD00001004159)
      - Dataset [EGAD00001006194](https://ega-archive.org/datasets/EGAD00001006194)
- Paper 2 (Martincorena 2018)
  - EGA Accessions:
    - EGAD00001004158
      - Study [EGAS00001001933](https://ega-archive.org/studies/EGAS00001001933)
      - Dataset [EGAD00001004158](https://ega-archive.org/datasets/EGAD00001004158)
    - EGAD00001004159
      - Study [EGAS00001002416](https://ega-archive.org/studies/EGAS00001002416)
      - Dataset [EGAD00001004159](https://ega-archive.org/datasets/EGAD00001004159)
- Paper 3 (Martincorena 2015)
  - EGA Accessions:
    - EGAS00001000860
      - Study [EGAS00001000860](https://ega-archive.org/studies/EGAS00001000860)
      - Dataset [EGAD00001001123](https://ega-archive.org/datasets/EGAD00001001123)
    - EGAS00001000515
      - Study [EGAS00001000515](https://ega-archive.org/studies/EGAS00001000515)
      - Dataset [EGAD00001001090](https://ega-archive.org/datasets/EGAD00001001090)
    - EGAS00001000603
      - Study [EGAS00001000603](https://ega-archive.org/studies/EGAS00001000603)
      - Dataset [EGAD00001000825](https://ega-archive.org/datasets/EGAD00001000825)
- Paper 4 (Lee-Six 2019)
  - EGA Accessions:
    - EGAD00001004192
      - Study [EGAS00001001855](https://ega-archive.org/studies/EGAS00001001855)
      - Study [EGAS00001002471](https://ega-archive.org/studies/EGAS00001002471)
      - Dataset [EGAD00001004192](https://ega-archive.org/datasets/EGAD00001004192)
    - EGAD00001004193
      - Study [EGAS00001001856](https://ega-archive.org/studies/EGAS00001001856)
      - Dataset [EGAD00001004193](https://ega-archive.org/datasets/EGAD00001004193)
- Paper 5 (Keogh 2018)
  - SRA Accessions:
    - SRP159015
      - [179 samples](https://www.ncbi.nlm.nih.gov/sra/?term=SRP159015)
      - Covaris sheared, PCR 6 cycles, 102 gene SureSelect probe enrich, PCR 10 cycles, HiSeq 2500 single lane, paired-end 2 Ã— 101 bp
- Paper 6 (Zhang 2019)
  - dbGaP Accessions:
    - [phs001808.v1.p1](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs001808.v1.p1)
- Paper 7 (Yokoyama 2019)
  - EGA Accessions:
    - EGAS00001003008
      - Study [EGAS00001003008](https://ega-archive.org/studies/EGAS00001003008)
      - Dataset [EGAD00001004464](https://ega-archive.org/datasets/EGAD00001004464)
    - EGAS00001003281
      - Study [EGAS00001003281](https://ega-archive.org/studies/EGAS00001003281)
      - Dataset [EGAD00001004462](https://ega-archive.org/datasets/EGAD00001004462)
    - EGAS00001003331
      - Study [EGAS00001003331](https://ega-archive.org/studies/EGAS00001003331)
      - Dataset [EGAD00010001631](https://ega-archive.org/datasets/EGAD00010001631)
- Paper 8 (Buscarlet 2017)
  - Not found any accession numbers, perhaps requires emailing corresponding author direct?
- Paper 9 (Yamaguchi 2022)
  - Would need a separate DAC data access request through EGA
- Paper 10 (Lawson 2020)
  - [EGAD00001006113](https://ega-archive.org/datasets/EGAD00001006113)
  - [EGAD00001006114](https://ega-archive.org/datasets/EGAD00001006114)
  - [EGAD00001006115](https://ega-archive.org/datasets/EGAD00001006115)
  - [EGAD00001006116](https://ega-archive.org/datasets/EGAD00001006116)
  - [EGAD00001006117](https://ega-archive.org/datasets/EGAD00001006117)
- Paper 11 (Brunner 2019)
  - [EGAD00001004578](https://ega-archive.org/datasets/EGAD00001004578)

### List of Dataset Accessions

This table is a duplicate-free list of the dataset accession numbers from the above papers. We could use the accession number as the dataset names in the code and manuscript, but perhaps more informative names would be better referring to tissue type etc, hence the placeholder Sample Name column. Current total data size is 27TB.

| Dataset UID | Orig Acc Numb| Our Sample Name | URL |Paper UID(s) | Data Access Control | Notes | Size (GB) |
|-------------|--------------------------|-----------------|-----|-------------|---------------------|-------|-----------|
| 1           | EGAD00001006969          | | [EGA](https://ega-archive.org/datasets/EGAD00001006969) | 1   | datasharing@sanger.ac.uk |63 files, CRAM, Illumina HiSeq 2500|14|
| 2           | EGAD00001004158          | | [EGA](https://ega-archive.org/datasets/EGAD00001004158) | 1,2 | datasharing@sanger.ac.uk |866 files, CRAM, Illumina HiSeq 2500|597|
| 3           | EGAD00001006194          | | [EGA](https://ega-archive.org/datasets/EGAD00001006194) | 1   | datasharing@sanger.ac.uk |1780 files, CRAM, HiSeq X Ten,Illumina HiSeq 2500,Illumina NovaSeq 6000|2653|
| 4           | EGAD00001004159          | | [EGA](https://ega-archive.org/datasets/EGAD00001004159) | 1,2 | datasharing@sanger.ac.uk |216 files, CRAM, HiSeq X Ten|1659|
| 5           | EGAD00001001123          | | [EGA](https://ega-archive.org/datasets/EGAD00001001123) | 3   | datasharing@sanger.ac.uk |16 files, CRAM, Illumina HiSeq 2000|214|
| 6           | EGAD00001001090          | | [EGA](https://ega-archive.org/datasets/EGAD00001001090) | 3   | datasharing@sanger.ac.uk |262 files, BAM, Illumina HiSeq 2000|204|
| 7           | EGAD00001000825          | | [EGA](https://ega-archive.org/datasets/EGAD00001000825) | 3   | datasharing@sanger.ac.uk |700 files, CRAM, Illumina HiSeq 2000|489|
| 8           | EGAD00001004192          | | [EGA](https://ega-archive.org/datasets/EGAD00001004192) | 4   | datasharing@sanger.ac.uk |2038 files, CRAM, HiSeq X Ten|10681|
| 9           | EGAD00001004193          | | [EGA](https://ega-archive.org/datasets/EGAD00001004193) | 4   | datasharing@sanger.ac.uk |1632 files, CRAM, Illumina HiSeq 2500|195|
| 10          | SRP159015                | | [SRA](https://www.ncbi.nlm.nih.gov/sra/?term=SRP159015) | 5   | open access?                    |179 files, SRA, Illumina HiSeq 2500|203|
| 11          | SRP219003                | | [phs001808.v1.p1](https://www.ncbi.nlm.nih.gov/projects/gap/cgi-bin/study.cgi?study_id=phs001808.v1.p1) [PRJNA531733](https://www.ncbi.nlm.nih.gov/bioproject/PRJNA531733) [SRP219003](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP219003)|6|JAAMH-DAC@list.nih.gov|152 files, SRA,Illumina HiSeq X Ten, HiSeq 2000|3103|
| 12          | EGAD00001004464          | | [EGA](https://ega-archive.org/datasets/EGAD00001004464) | 7   | Akira Yokoyama kyototumorpatho-labo@umin.ac.jp|523 files, BAM, Illumina HiSeq 2000|5950|
| 13          | EGAD00001004462          | | [EGA](https://ega-archive.org/datasets/EGAD00001004462) | 7   | Akira Yokoyama kyototumorpatho-labo@umin.ac.jp|20 files, BAM, Illumina NovaSeq 6000|1209|
| 14          | EGAD00010001631          | | [EGA](https://ega-archive.org/datasets/EGAD00010001631) | 7   | Akira Yokoyama kyototumorpatho-labo@umin.ac.jp|124 files, "raw" format, GeneChip Human Mapping 250K NspI|5|
